src/box/gc.c

```diff
@@ -903,24 +903,30 @@ gc_erase_consumer_on_commit(struct trigger *trigger, void *event)
 int
 gc_erase_consumer(const struct tt_uuid *uuid)
 {
-	assert(in_txn() != NULL);
 	struct gc_consumer *consumer = gc_consumer_by_uuid(uuid);
 	if (!gc_schema_supports_persistent_consumers()) {
 		if (consumer != NULL)
 			consumer->is_persistent = false;
 		return 0;
 	}
 
+	bool is_autocommit = in_txn() == NULL;
+	if (is_autocommit && txn_begin() == NULL)
+		return -1;
+
 	if (boxk(IPROTO_DELETE, BOX_GC_CONSUMERS_ID, "[%s]",
-		 tt_uuid_str(uuid)) != 0)
+		 tt_uuid_str(uuid)) != 0) {
+		if (is_autocommit)
+			txn_abort(in_txn());
 		return -1;
+	}
 
 	/* Do actual work on commit. */
 	struct trigger *on_commit =
 		xregion_alloc_object(&in_txn()->region, struct trigger);
 	trigger_create(on_commit, gc_erase_consumer_on_commit, consumer, NULL);
-	txn_stmt_on_commit(txn_current_stmt(in_txn()), on_commit);
-	return 0;
+	txn_stmt_on_commit(txn_last_stmt(in_txn()), on_commit);
```

Comment:
Added an explanation to the commit message:
```
In order to call `gc_erase_consumer` under the hood, the commit
populates it with autocommit mode so that it can be called without an
active transaction. However, underlying `txn_current_stmt` call is not
appropriate in this case - current statement is actually one that is
being executed: for example, we used to call this function in
`_cluster:on_replace` - in this case current statement is the replace in
`_cluster`. Since we don't have current statement in autocommit mode,
let's use last statement (the actual deletion in `_gc_consumers`) instead.
```